File: java/app/src/main/java/com/jossy/android/mobilebenchmarkappjava/activity/BaseTestActivity.java
package com.jossy.android.mobilebenchmarkappjava.activity;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.MutableLiveData;

import com.jossy.android.mobilebenchmarkappjava.BenchmarkApplication;
import com.jossy.android.mobilebenchmarkappjava.R;
import com.jossy.android.mobilebenchmarkappjava.consts.Config;
import com.jossy.android.mobilebenchmarkappjava.data.TestEntry;
import com.jossy.android.mobilebenchmarkappjava.data.TestResult;
import com.jossy.android.mobilebenchmarkappjava.io.BufferedCsvWriter;

import java.io.File;
import java.util.concurrent.Executors;

/**
 * Base class for test activities to reduce code duplication.
 * Handles CSV writing, result logging, and progress updates.
 */
public abstract class BaseTestActivity extends AppCompatActivity {

    private static final String TAG = "BaseTestActivity";

    protected TextView statusText;
    protected final MutableLiveData<Integer> progressLiveData = new MutableLiveData<>();
    protected BufferedCsvWriter csvWriter;
    protected long suiteStartTime;
    protected String csvPath;
    protected int currentSampleIndex = 0;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        initializeActivity();

        progressLiveData.observe(this, this::updateProgressText);

        csvPath = getIntent().getStringExtra("csv_path");
        Log.d(TAG, "Test activity initialized");

        startBenchmarkAsync();
    }

    /**
     * Subclasses implement this to set content view and initialize UI elements.
     */
    protected abstract void initializeActivity();

    /**
     * Subclasses implement this to perform the actual benchmark.
     */
    protected abstract void executeBenchmark() throws Exception;

    /**
     * Called when benchmark results need to be formatted for display.
     */
    protected abstract String getProgressDisplayText(int currentIteration);

    /**
     * Start benchmark execution on background thread.
     */
    protected void startBenchmarkAsync() {
        Executors.newSingleThreadExecutor().execute(this::runBenchmarkTask);
    }

    /**
     * Main benchmark task - wraps execution with error handling.
     */
    private void runBenchmarkTask() {
        suiteStartTime = System.currentTimeMillis();
        try {
            executeBenchmark();
            handleSuccess(System.currentTimeMillis() - suiteStartTime);
        } catch (Exception e) {
            handleError(e);
        }
    }

    /**
     * Log a single test iteration result to CSV.
     */
    protected void logTestResult(int iteration, TestResult result) {
        try {
            if (csvWriter == null) {
                initializeCsvWriter();
            }

            long start = System.currentTimeMillis();
            long duration = result.getExecutionTimeMs();
            long elapsed = System.currentTimeMillis() - suiteStartTime;

            TestEntry entry = new TestEntry(iteration, result, start, duration, elapsed);
            csvWriter.write(entry);
        } catch (Exception e) {
            Log.e(TAG, "Failed to log result: " + e.getMessage(), e);
        }
    }

    /**
     * Initialize CSV writer with default settings.
     */
    protected void initializeCsvWriter() throws Exception {
        if (csvWriter != null)
            return;

        File file = new File(csvPath != null ? csvPath : getDefaultCsvPath());
        csvWriter = new BufferedCsvWriter(file, Config.bufferSize, 64 * 1024);
        csvWriter.initialize();
    }

    /**
     * Get default CSV file path if not provided.
     */
    protected String getDefaultCsvPath() {
        return getFilesDir() + "/benchmark_results.csv";
    }

    /**
     * Update progress display on UI thread.
     */
    protected void updateProgress(int currentIteration) {
        progressLiveData.postValue(currentIteration);
    }

    /**
     * Update progress text display.
     */
    private void updateProgressText(Integer currentIteration) {
        if (statusText != null) {
            statusText.setText(getProgressDisplayText(currentIteration));
        }
    }

    /**
     * Handle successful benchmark completion.
     */
    protected void handleSuccess(long totalTime) {
        Log.i(TAG, "Benchmark completed in " + totalTime + "ms");
        runOnUiThread(() -> {
            TestResult result = new TestResult(
                    getTestName(),
                    totalTime,
                    "Batch completed: " + Config.sampleCount + " samples",
                    true);
            closeCsvWriter();
            returnResult(result);
        });
    }

    /**
     * Handle benchmark error.
     */
    protected void handleError(Exception e) {
        Log.e(TAG, "Benchmark failed: " + e.getMessage(), e);
        runOnUiThread(() -> {
            TestResult result = new TestResult(
                    getTestName(),
                    0,
                    "Error: " + e.getMessage(),
                    false);
            closeCsvWriter();
            returnResult(result);
        });
    }

    /**
     * Close CSV writer safely.
     */
    protected void closeCsvWriter() {
        try {
            if (csvWriter != null) {
                csvWriter.flush();
                csvWriter.close();
            }
        } catch (Exception e) {
            Log.e(TAG, "Error closing CSV writer: " + e.getMessage(), e);
        }
    }

    /**
     * Return test result to calling activity.
     */
    private void returnResult(TestResult result) {
        Intent intent = new Intent();
        intent.putExtra(BenchmarkApplication.RESULT, result);
        setResult(RESULT_OK, intent);
        finish();
    }

    /**
     * Get test name for logging and results.
     */
    protected abstract String getTestName();
}

File: kotlin/app/src/main/kotlin/com/jossy/android/mobilebenchmarkappjava/activity/BaseTestActivity.kt
package com.jossy.android.mobilebenchmarkappjava.activity

import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.util.Log
import android.widget.LinearLayout
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.MutableLiveData
import com.jossy.android.mobilebenchmarkappjava.BenchmarkApplication
import com.jossy.android.mobilebenchmarkappjava.consts.Config
import com.jossy.android.mobilebenchmarkappjava.data.TestEntry
import com.jossy.android.mobilebenchmarkappjava.data.TestResult
import com.jossy.android.mobilebenchmarkappjava.io.BufferedCsvWriter
import java.io.File
import java.util.concurrent.Executors

abstract class BaseTestActivity : AppCompatActivity() {

    companion object {
        private const val TAG = "BaseTestActivity"
    }

    protected lateinit var statusText: TextView
    protected val progressLiveData = MutableLiveData<Int>()
    protected var csvWriter: BufferedCsvWriter? = null
    protected var suiteStartTime = 0L
    protected var csvPath: String? = null
    protected var currentSampleIndex = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        initializeActivity()

        progressLiveData.observe(this) { updateProgressText(it) }

        csvPath = intent.getStringExtra("csv_path")
        Log.d(TAG, "Test activity initialized")

        startBenchmarkAsync()
    }

    protected abstract fun initializeActivity()

    protected abstract fun executeBenchmark()

    protected abstract fun getProgressDisplayText(currentIteration: Int): String

    protected abstract fun getTestName(): String

    protected fun startBenchmarkAsync() {
        Executors.newSingleThreadExecutor().execute(this::runBenchmarkTask)
    }

    private fun runBenchmarkTask() {
        suiteStartTime = System.currentTimeMillis()
        try {
            executeBenchmark()
            handleSuccess(System.currentTimeMillis() - suiteStartTime)
        } catch (e: Exception) {
            handleError(e)
        }
    }

    protected fun logTestResult(iteration: Int, result: TestResult) {
        try {
            if (csvWriter == null) {
                initializeCsvWriter()
            }

            val start = System.currentTimeMillis()
            val duration = result.executionTimeMs
            val elapsed = System.currentTimeMillis() - suiteStartTime

            val entry = TestEntry(iteration, result, start, duration, elapsed)
            csvWriter?.write(entry)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to log result: ${e.message}", e)
        }
    }

    protected fun initializeCsvWriter() {
        if (csvWriter != null) return

        val file = File(csvPath ?: getDefaultCsvPath())
        csvWriter = BufferedCsvWriter(file, 1000, 64 * 1024).apply { initialize() }
    }

    protected fun getDefaultCsvPath(): String =
        "${filesDir.absolutePath}/benchmark_results.csv"

    protected fun updateProgress(currentIteration: Int) {
        progressLiveData.postValue(currentIteration)
    }

    private fun updateProgressText(currentIteration: Int) {
        statusText.text = getProgressDisplayText(currentIteration)
    }

    protected fun handleSuccess(totalTime: Long) {
        Log.i(TAG, "Benchmark completed in ${totalTime}ms")
        runOnUiThread {
            val result = TestResult(
                testName = getTestName(),
                executionTimeMs = totalTime,
                details = "Batch completed: ${Config.sampleCount} samples",
                success = true
            )
            closeCsvWriter()
            returnResult(result)
        }
    }

    protected fun handleError(e: Exception) {
        Log.e(TAG, "Benchmark failed: ${e.message}", e)
        runOnUiThread {
            val result = TestResult(
                testName = getTestName(),
                executionTimeMs = 0,
                details = "Error: ${e.message}",
                success = false
            )
            closeCsvWriter()
            returnResult(result)
        }
    }

    protected fun closeCsvWriter() {
        try {
            csvWriter?.apply {
                flush()
                close()
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error closing CSV writer: ${e.message}", e)
        }
    }

    private fun returnResult(result: TestResult) {
        Intent().apply {
            putExtra(BenchmarkApplication.RESULT, result)
            setResult(RESULT_OK, this)
        }
        finish()
    }
}
